treeherder/services/taskcluster.py

```diff
@@ -0,0 +1,156 @@
+import logging
+from typing import List
+
+import jsone
+import taskcluster
+
+from treeherder.utils.taskcluster_lib_scopes import satisfiesExpression
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_ROOT_URL = 'https://firefox-ci-tc.services.mozilla.com'
+
+
+class TaskclusterModel:
+    """Javascript -> Python rewrite of frontend's TaskclusterModel"""
+
+    def __init__(self, root_url, client_id=None, access_token=None):
+        options = {'rootUrl': root_url}
+        credentials = {}
+
+        if client_id:
+            credentials['clientId'] = client_id
+        if access_token:
+            credentials['accessToken'] = access_token
+
+        # Taskcluster APIs
+        self.hooks = taskcluster.Hooks({**options, 'credentials': credentials})
+
+        # Following least-privilege principle, as services
+        # bellow don't really need authorization credentials.
+        self.queue = taskcluster.Queue(options)
+        self.auth = taskcluster.Auth(options)
+
+    def set_root_url(self, root_url):
+        for service in (self.hooks, self.queue, self.auth):
+            service.options['rootUrl'] = root_url
+
+    def trigger_action(self, action, task_id, decision_task_id, input, root_url=None) -> str:
+        if root_url is not None:
+            self.set_root_url(root_url)
+
+        actions_context = self._load(decision_task_id, task_id)
+        action_to_trigger = self._get_action(actions_context['actions'], action)
+
+        return self._submit(
+            action=action_to_trigger,
+            decision_task_id=decision_task_id,
+            task_id=task_id,
+            input=input,
+            static_action_variables=actions_context['staticActionVariables']
+        )
+
+    def _load(self, decision_task_id: str, task_id: str) -> dict:
+        if not decision_task_id:
+            raise ValueError("No decision task, can't find taskcluster actions")
+
+        # fetch
+        logger.debug('Fetching actions.json...')
+        actions_json = self.queue.getLatestArtifact(decision_task_id, 'public/actions.json')
+        task_definition = self.queue.task(task_id)
+
+        if actions_json['version'] != 1:
```

Comment:
why does this need to == 1?  is there a reference to what is required there?  what if it is 1.1 or 2 ?  Please add a comment to indicate what is needed for version=1

Reply:
There is a formal specification for why this is. It's available [here](https://docs.taskcluster.net/docs/manual/design/conventions/actions/spec#formal-specification).
According to that, the `version` is a mandatory field which can only hold a number, from an enum list. For some time now, that enum list has only had one element and that is `1`.

I guess this is support work, in case Taskcluster team prepares **big** & API-breaking changes. So far, they didn't need to use this field.

Reply:
could you add a comment pointing to the definition?

Reply:
I'm going to take care of this in a separate bug.